rust/kernel/of.rs

```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Devicetree and Open Firmware abstractions.
+//!
+//! C header: [`include/linux/of_*.h`](../../../../include/linux/of_*.h)
+
+use alloc::boxed::Box;
+
+use crate::{
+    bindings, c_types,
+    error::{Error, Result},
+    types::PointerWrapper,
+    CStr,
+};
+
+use core::mem::transmute;
+
+type InnerTable = Box<[bindings::of_device_id; 2]>;
+
+/// Wraps a kernel Open Firmware / devicetree match table.
+///
+/// Rust drivers may create this structure to match against devices
+/// described in the devicetree.
+///
+/// The ['PointerWrapper'] trait provides conversion to/from a raw pointer,
+/// suitable to be assigned to a `bindings::device_driver::of_match_table`.
+///
+/// # Invariants
+///
+/// The final array element is always filled with zeros (the default).
+pub struct OfMatchTable(InnerTable);
+
+impl OfMatchTable {
+    /// Creates a [`OfMatchTable`] from a single `compatible` string.
+    pub fn new(compatible: &CStr<'static>) -> Result<Self> {
+        let tbl: InnerTable = Box::try_new([
```

Comment:
nit: do we need the type (`InnerTable`) explicitly given here?

Reply:
You are right - there's no need. I don't like unnecessary explicit type annotations either. I'll do a follow-up PR to eliminate it.